headers/storage/buffer.h

```diff
@@ -0,0 +1,80 @@
+//
+// Created by Sameer Raj on 13/07/24.
+//
+
+#ifndef BUFFER_H
+#define BUFFER_H
+
+#include <atomic>
+#include <unordered_map>
+
+#include "../configs/constants.h"
+#include "block.h"
+#include "lock.h"
+
+using BufferId = unsigned int;
+// 128 MB for 8KB buffer
+
+struct BufferTag {
+    BlockNumber blockNumber;
+    uint64_t fileNumber;
+};
+
+struct BufferDescriptor {
+    BufferDescriptor(const BufferTag &tag, BufferId id)
+        : tag(tag),
+          id(id),
+          pinCount(0),
+          usageCount(0),
+          isDirty(false),
+          contentLock() {
+    }
+
+    BufferTag tag;
+    BufferId id;
+    std::atomic<unsigned int> pinCount;
+    std::atomic<unsigned int> usageCount;
+    std::atomic<bool> isDirty;
```

Comment:
Does postgres also use atomic variables fro these counters and flags, it makes sense to do do but just need to know if postgres does?

Reply:
Postgres uses a single atomic variable: pg_atomic_uint32 state;
Different bits of state stores different values using:
#define BUF_STATE_GET_REFCOUNT(state) ((state) & BUF_REFCOUNT_MASK)
#define BUF_REFCOUNT_MASK ((1U << 18) - 1)

in short: yes